File: algorithm/baekjoon/codeplus/first-solve/src/main/java/com/keunwon/algorithm/dfs/Problem2668.java
package com.keunwon.algorithm.dfs;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Scanner;

public class Problem2668 {

    static int N;
    static int[] numbers;
    static List<Integer> result = new ArrayList<>();

    public static void main(String[] args) {
        input();

        for (int i = 1; i <= N; i++) {
            dfs(i, i, new boolean[N + 1]);
        }

        Collections.sort(result);

        System.out.println(result.size());
        for (Integer value : result) {
            System.out.println(value);
        }
    }

    static void dfs(int x, int target, boolean[] visited) {
        if (target == numbers[x]) {
            result.add(target);
        }

        visited[x]  = true;

        if (!visited[numbers[x]]) {
            dfs(numbers[x], target, visited);
        }
    }

    static void input() {
        Scanner sc = new Scanner(System.in);

        N = sc.nextInt();
        numbers = new int[N + 1];

        for (int i = 1; i <= N; i++) {
            numbers[i] = sc.nextInt();
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem2668.kt
package com.keunwon.algorithm.baekjoon

/**
 * <p>
 * 이름: 숫자고르기
 * 난이도: 골드-5
 * </p>
 *
 * @see <a href="https://www.acmicpc.net/problem/2668">숫자고르기 (골드-5)</a>
 **/
class Problem2668 {
    private lateinit var numbers: IntArray
    private lateinit var visited: BooleanArray

    private var result = mutableListOf<Int>()

    fun solution(numbers: IntArray): IntArray {
        this.numbers = intArrayOf(0, *numbers)
        this.visited = BooleanArray(numbers.size + 1)

        for (i in 1..numbers.size) {
            dfs(i, i)
        }
        result.add(0, result.size)
        return result.toIntArray()
    }

    private fun dfs(index: Int, num: Int) {
        if (numbers[index] == num) {
            result.add(num)
            return
        }

        if (!visited[index]) {
            visited[index] = true
            dfs(numbers[index], num)
            visited[index] = false
        }
    }
}

fun main() {
    val br = System.`in`.bufferedReader()
    val bw = System.out.bufferedWriter()

    val n = br.readLine().toInt()
    val numbers = IntArray(n) { br.readLine().toInt() }
    Problem2668().solution(numbers).forEach {
        bw.write("$it")
        bw.newLine()
    }

    bw.flush()
    bw.close()
    br.close()
}
